Add standalone `deriving`

sarajuhosova opened the pull request:
This PR adds a way to derive type class instances which are available on the Agda side.
This is done by postulating the instance (there is no special pragma needed).

Agda:
```agda
data Optional (a : Set) : Set where
  Of    : a → Optional a
  Empty : Optional a
{-# COMPILE AGDA2HS Optional #-}

postulate instance iOptionalEq : ⦃ Eq a ⦄ → Eq (Optional a)
{-# COMPILE AGDA2HS iOptionalEq #-}
```

Haskell:
```haskell
data Optional a = Of a
                | Empty

deriving instance (Eq a) => Eq (Optional a)
```

## Deriving Strategies

It is also possible to specify a deriving strategy in the pragma.

NOTE: **deriving via** is not included, since that's quite a bit more complex.

Agda:
```agda
postulate instance iPlanetShow : Show Planet

{-# COMPILE AGDA2HS iPlanetShow derive stock #-}
```

Haskell:
```haskell
deriving stock instance Show Planet
```

flupe commented on src/Agda2Hs/Pragma.hs:
```diff
@@ -74,12 +81,13 @@ processPragma :: QName -> C ParsedPragma
+    | s == "unboxed-strict"      -> return $ UnboxPragma Strict
+    | s == "transparent"         -> return TransparentPragma
+    | s == newtypePragma         -> return $ NewTypePragma []
+    | "strategy:" `isPrefixOf` s -> return $ DerivePragma (parseStrategy (drop (length "strategy:") s))
```
I'm in favor of dropping the `strategy:` (do you even check "strategy" is written?) as the fact you're compiling an instance makes it redundant.

sarajuhosova replied:
I've changed this to be included as part of the `derive` clause.

flupe reviewed (commented):
Looking nice!

I'm wondering if it makes sense to generate standalone deriving declarations *only* for *postulated* instances. In particular, if you define a wrapper record (on the Agda side) that will be compiled as a `newtype`, it is very likely you can define a proper instance yourself, even if the Haskell instance ought to be derived automatically.

jespercockx commented on docs/source/features.md:
````diff
@@ -407,7 +409,7 @@ class Ord a where
 ```
 
-## Coppaterns
+### Coppaterns in Type Class Instances
````
This is not your fault, but it should be "copatterns" (or "co-patterns") rather than coppatterns. No cops allowed here!

jespercockx reviewed (approved):
Looks good to me. This is a great improvement to dealing with `deriving`!

sarajuhosova commented on src/Agda2Hs/Compile/ClassInstance.hs:
```diff
@@ -42,16 +43,21 @@ enableCopatterns = local $ \e -> e { copatternsEnabled = True }
+compileInstance :: InstanceTarget -> Definition -> C (Hs.Decl ())
+compileInstance ToDerivation def@Defn{..} = do
+  tellExtension StandaloneDeriving
   ir <- compileInstRule [] (unEl defType)
```
Added!

sarajuhosova commented:
@lucas, I've added support for compiling implemented as well as postulated instances into a `deriving` clause. This can now be done with the `derive` pragma, which has also replaced the `strategy` annotation.

flupe reviewed (approved):
Looks good to me, well done!
